README.md
A Dart implementation of Java's [EnumSet](https://docs.oracle.com/javase/8/docs/api/java/util/EnumSet.html) (or bit fields).

Stores values by assigning a power-of-two value to each one and doing bitwise operations on integers. 

```dart
import "package:enumset/enumset.dart";

enum Suits {spades, clubs, hearts, diamonds}

void main() {
  final EnumSet<Suits> reds = EnumSet(Suits.values);
  // Suits.spades   = 0001 = 1
  // Suits.clubs    = 0010 = 2
  // Suits.hearts   = 0100 = 4
  // Suits.diamonds = 1000 = 8

  reds.addAll([Suits.hearts, Suits.diamonds]);
  print(reds);  // "Suits 1100"
  // Adds the bits of hearts and diamonds together: 
  //
  //   0100 = 4 (hearts)
  // | 1000 = 8 (diamonds)
  // ------
  //   1100 = 12 (hearts + diamonds)

  print(reds.contains(Suits.spades));  // false
  // Performs an AND operation:
  //
  //   1100 = 12 (hearts + diamonds)
  // & 0001 = 1 (spades)
  // ------
  //   0000 = 0 (false)

  print(reds.contains(Suits.hearts));  // true
  //   1100 = 12 (hearts + diamonds)
  // & 0100 = 4 (hearts)
  // ------
  //   0100 = 4 (hearts)

  reds.remove(Suits.hearts);
  // performs AND with NOT: 
  // 
  // ~0100 = 1011 (not hearts)
  // 
  //   1100 = 12 (hearts + diamonds)
  // & 1011 = 11 (-hearts)
  // ------
  //   1000 = 8 (diamonds)
}
```

## Usage

Use this as you would a `Set` of enum values. The only difference is you must first create it by passing in all possible values.

```dart
import "package:enumset/enumset.dart";

enum Suits {spades, clubs, hearts, diamonds}

void main() {
  final EnumSet<Suits> reds = EnumSet(Suits.values);  // empty
  print(reds.contains(Suits.spades));  // false
  print(reds.contains(Suits.hearts));  // false (it's empty)

  reds.addAll([Suits.hearts, Suits.diamonds]);
  print(reds.contains(Suits.spades));  // false
  print(reds.contains(Suits.hearts));  // true

  // Of course, you could also do this:
  final Set<Suits> blacks = {Suits.spades, Suits.clubs};
  print(blacks.contains(Suits.spades));  // true
  print(blacks.contains(Suits.hearts));  // false
}
```

You should probably just use `Set`. This is just for people interested in bit fields, since it's technically more memory-efficient.

CHANGELOG.md
## 1.0.0 

- First release

## 0.0.1

- Initial version.


test/enumset_test.dart
import "package:enumset/enumset.dart";
import "package:test/test.dart";

enum Suits {spades, clubs, hearts, diamonds}
enum Colors {red, green, blue, yellow}

void main() {
  test("toString() test", () {
    final EnumSet<Suits> reds = EnumSet(Suits.values);
    expect(reds.toString(), equals("Suits 0"));
    expect(reds.copy().toString(), equals("Suits 0"));

    reds.add(Suits.spades);
    expect(reds.toString(), equals("Suits 1"));

    reds..add(Suits.hearts)..add(Suits.diamonds);
    expect(reds.toString(), equals("Suits 1101"));
  });

  test("Equality", () {
    // Empty sets are equal
    final EnumSet<Suits> reds = EnumSet(Suits.values);
    final EnumSet<Suits> blacks = EnumSet(Suits.values);
    final EnumSet<Suits> reversed = EnumSet([
      Suits.diamonds, Suits.hearts, Suits.clubs, Suits.spades,
    ]);
    expect(reds.copy(), equals(reds));  // copy
    expect(reds, equals(blacks));  // equal sets
    expect(reds, equals(reds));  // identical
    expect(reds, isNot(equals(reversed)));  // order matters

    // Adding elements makes a difference
    reds.addAll([Suits.hearts, Suits.diamonds]);
    blacks.addAll([Suits.spades, Suits.clubs]);
    expect(reds, isNot(equals(blacks)));

    // Manual creation does not matter.
    final EnumSet<Suits> reds2 = EnumSet(Suits.values)
      ..addAll([Suits.hearts, Suits.diamonds]);
    expect(reds, equals(reds2));
  });

  test("Operator []", () {
    final EnumSet<Suits> suits1 = EnumSet(Suits.values);
    expect(suits1[Suits.spades], equals(1));
    expect(suits1[Suits.clubs], equals(2));
    expect(suits1[Suits.hearts], equals(4));  
    expect(suits1[Suits.diamonds], equals(8));

    final EnumSet<Suits> suits2 = EnumSet([
      Suits.diamonds, Suits.hearts, Suits.clubs, Suits.spades,
    ]);
    expect(suits2[Suits.spades], equals(8));
    expect(suits2[Suits.clubs], equals(4));
    expect(suits2[Suits.hearts], equals(2));  
    expect(suits2[Suits.diamonds], equals(1));
  });

  test("Hashing", () {
    final EnumSet<Suits> reds = EnumSet(Suits.values)
      ..addAll([Suits.hearts, Suits.diamonds]);
    final EnumSet<Suits> reds2 = reds.copy();
    final EnumSet<Suits> blacks = EnumSet(Suits.values)
      ..addAll([Suits.spades, Suits.clubs]);
    final EnumSet<Colors> colors = EnumSet(Colors.values)
      ..addAll([Colors.red, Colors.green]);  // same bit value
    final Map<EnumSet, String> sets = {
      reds: "Should not appear",
      reds2: "reds",  // equal to reds
      blacks: "blacks",  // same flags as reds
      colors: "colors",  // same bits as blacks
    };

    expect(colors.bits, equals(blacks.bits));
    expect(sets[reds], equals("reds"));
    expect(sets[reds2], equals("reds"));
    expect(sets[blacks], equals("blacks"));
    expect(sets[colors], equals("colors"));

    expect(reds.hashCode, equals(reds2.hashCode));
    expect(reds.hashCode, isNot(equals(colors.hashCode)));

    // Equal flags, different hash codes
    expect(reds.flags, equals(blacks.flags));
    expect(reds.hashCode, isNot(equals(blacks.hashCode)));
    expect(reds, isNot(equals(blacks)));

    // Equal bits, different hash codes
    expect(blacks.bits, equals(colors.bits));
    expect(blacks.hashCode, isNot(equals(colors.hashCode)));
    expect(blacks, isNot(equals(colors)));
  });

  test("Add all", () {
    final EnumSet<Suits> reds = EnumSet(Suits.values);
    final EnumSet<Suits> reds2 = reds.copy();

    reds.addAll([Suits.hearts, Suits.diamonds]);
    reds2..add(Suits.hearts)..add(Suits.diamonds);
    expect(reds, equals(reds2));
  });

  test("Add and remove", () {
    final EnumSet<Suits> reds = EnumSet(Suits.values);

    expect(reds.bits, equals(0));  // empty

    reds.add(Suits.hearts);
    expect(reds.bits, equals(4));  // add hearts = 4

    reds.add(Suits.spades);
    expect(reds.bits, equals(5));  // add spades = 1

    reds.remove(Suits.spades);
    expect(reds.bits, equals(4));  // remove spades = 1

    reds.remove(Suits.clubs);
    expect(reds.bits, equals(4));  // remove clubs (not in set)

    reds.remove(Suits.hearts);
    expect(reds.bits, equals(0));  // remove hearts = 4

    reds.remove(Suits.hearts);
    expect(reds.bits, equals(0));  // remove hearts (not in set)
  });

  test("Contains", () {
    final EnumSet<Suits> reds = EnumSet(Suits.values);

    expect(reds.contains(Suits.spades), isFalse);
    expect(reds.contains(Suits.hearts), isFalse);

    reds.add(Suits.hearts);
    expect(reds.contains(Suits.spades), isFalse);
    expect(reds.contains(Suits.hearts), isTrue);  // added

    reds.remove(Suits.hearts);
    expect(reds.contains(Suits.spades), isFalse);
    expect(reds.contains(Suits.hearts), isFalse);  // removed
  });

  test("toList()", () {
    final EnumSet<Suits> reds = EnumSet(Suits.values);
    expect(reds.toSet(), isEmpty);

    reds.add(Suits.hearts);
    expect(reds.toSet(), equals({Suits.hearts}));

    reds.add(Suits.diamonds);
    expect(reds.toSet(), equals({Suits.diamonds, Suits.hearts}));
    expect(reds.toSet(), equals({Suits.hearts, Suits.diamonds}));

    reds.remove(Suits.diamonds);
    expect(reds.toSet(), equals({Suits.hearts}));

    reds.remove(Suits.hearts);
    expect(reds.toSet(), isEmpty);
  });
}


lib/enumset.dart
library enumset;

/// A bit field representing a set of enum instances. See Java's EnumSet.
/// 
/// This class works by assigning an increasing power of two to each instance 
/// of the enum, in the order it was declared, and saving it to [flags]. 
/// 
/// ```dart
/// enum Suit {spades, clubs, hearts, diamonds}
/// final EnumSet<Suit> reds = EnumSet(Suit.values)
/// print(reds);  // 0, because it's empty
/// print(reds.flags); 
/// // {Suit.spades: 1, Suit.clubs: 2, Suit.hearts: 4, Suit.diamonds: 8}
/// ```
/// 
/// This is done so that the entire collection can be represented in binary.
/// 
/// ```dart
/// reds.addAll(Suit.hearts, Suit.diamonds);
/// // addAll() adds the bits of each element together
/// // So, 4 + 8 = 12. In binary: 
/// //   0100
/// // + 1000
/// // -------
/// //   1100
/// print(reds);  // 1100 = 12
/// ```
/// 
/// Adding is equal to bitwise OR, and removing elements is bitwise XOR.
/// Checking if an element is in the set is done with bitwise AND.
/// 
/// ```dart
/// print(reds.contains(Suits.spades));  // false
/// //   1100
/// // & 0001
/// // ------
/// //   0000
/// 
/// print(reds.contains(Suits.hearts));  // true
/// //   1100
/// // & 0100
/// // ------
/// //   0100, which is not 0 (and is equal to 4 -- the value of Suit.hearts)
/// ```
/// 
/// You can also convert an EnumSet to a list of its values: 
/// 
/// ```dart
/// print(reds.toList());  // [Suit.hearts, Suit.diamonds]
/// ```
class EnumSet<T> {
  /// Maps instances of [T] to their power-of-two values.
  Map<T, int> flags = {};

  /// The bit field this set represents.
  int bits = 0;

  /// Creates an empty bit field for [values].
  EnumSet(List<T> values) : flags = {
    for (int index = 0; index < values.length; index++)
      values [index]: 1 << index,
  };

  /// Used for copying.
  EnumSet._(this.flags, this.bits);

  @override
  String toString() => "$T ${bits.toRadixString(2)}";

  @override
  bool operator ==(Object other) => other is EnumSet<T> 
    && flags.keys.every((key) => other.flags[key] == flags[key])
    && other.bits == bits;

  @override
  int get hashCode => bits.hashCode + flags.hashCode;

  /// Gets the binary value of the passed in value.
  /// 
  /// Since this class is optimized for enums, this will throw if you pass in a
  /// value that wasn't included when creating this set. This is because enums
  /// have a `.values` list that are guaranteed to contain all the elements of 
  /// the enum.
  int operator [](T value) => flags[value]!;

  /// Copies this set so any modifications made on one don't affect the other.
  EnumSet<T> copy() => EnumSet._(flags, bits);

  /// Adds all the given values to this set.
  void addAll(Iterable<T> values) => values.forEach(add);
    
  /// Adds the given value to this set.
  void add(T value) => bits |= this[value];

  /// Removes the given value from this set.
  void remove(T value) => bits &= ~this[value];

  /// Checks if the given value is contained within this set.
  bool contains(T value) => bits & this[value] != 0;

  /// Returns all the elements of this set.
  Set<T> toSet() => {
    for (final T value in flags.keys)
      if (contains(value)) value
  };
} 


example/enumset_example.dart
// ignore_for_file: avoid_print

import "package:enumset/enumset.dart";

enum Suits {spades, clubs, hearts, diamonds}

void main() {
  final EnumSet<Suits> reds = EnumSet(Suits.values)
    ..addAll([Suits.hearts, Suits.diamonds]);
  // Suits.spades   = 0001 = 1
  // Suits.clubs    = 0010 = 2
  // Suits.hearts   = 0100 = 4
  // Suits.diamonds = 1000 = 8

  print(reds);  // "Suits 1100"
  // Adds the bits of hearts and diamonds together: 
  //
  //   0100 = 4 (hearts)
  // | 1000 = 8 (diamonds)
  // ------
  //   1100 = 12 (hearts + diamonds)

  print(reds.contains(Suits.spades));  // false
  // Performs an AND operation:
  //
  //   1100 = 12 (hearts + diamonds)
  // & 0001 = 1 (spades)
  // ------
  //   0000 = 0 (false)

  print(reds.contains(Suits.hearts));  // true
  //   1100 = 12 (hearts + diamonds)
  // & 0100 = 4 (hearts)
  // ------
  //   0100 = 4 (hearts)

  reds.remove(Suits.hearts);
  // performs AND with NOT: 
  // 
  // ~0100 = 1011 (not hearts)
  // 
  //   1100 = 12 (hearts + diamonds)
  // & 1011 = 11 (-hearts)
  // ------
  //   1000 = 8 (diamonds)
}
